runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/impl/StandardRetryStrategy.kt

```diff
@@ -0,0 +1,160 @@
+     * @param fromToken A [RetryToken] which grants the strategy capacity to execute a try. This token is resolved
+     * inside the function by calling [notifySuccess][RetryToken.notifySuccess],
+     * [notifyFailure][RetryToken.notifyFailure], or [scheduleRetry][RetryToken.scheduleRetry].
+     * @param previousResult The [Result] from the prior loop iteration. This is used in the case of a timeout to
+     * include in the thrown exception.
+     */
+    private tailrec suspend fun <R> doTryLoop(
+        block: suspend () -> R,
+        policy: RetryPolicy<R>,
+        attempt: Int,
+        fromToken: RetryToken,
+        previousResult: Result<R>?,
+    ): R {
+        val callResult = runCatching { block() }
+        if (callResult.exceptionOrNull() is TimeoutCancellationException) {
+            throwTimeOut(fromToken, attempt, previousResult)
+        }
+
+        val nextToken = try {
+            when (val evaluation = policy.evaluate(callResult)) {
+                is RetryDirective.TerminateAndSucceed ->
+                    return success(fromToken, callResult)
+
+                is RetryDirective.TerminateAndFail ->
+                    throwFailure(fromToken, attempt, callResult)
+
+                is RetryDirective.RetryError ->
+                    if (attempt >= options.maxAttempts) {
+                        throwTooManyAttempts(fromToken, attempt, callResult)
+                    } else {
+                        // Prep for another loop
+                        backoffDelayer.backoff(attempt)
+                        fromToken.scheduleRetry(evaluation.reason)
+                    }
+            }
+        } catch (e: TimeoutCancellationException) {
+            throwTimeOut(fromToken, attempt, callResult)
+        }
+
+        return doTryLoop(block, policy, attempt + 1, nextToken, callResult)
+    }
+
+    /**
+     * Handles the successful termination of the retry loop by marking the [RetryToken] as successful and getting the
+     * [Result]'s value.
+     * @param token The [RetryToken] used in the attempt that was successful.
+     * @param result The [Result] that was evaluated to be successful.
+     * @return The [Result]'s value.
+     */
+    private suspend fun <R> success(token: RetryToken, result: Result<R>): R {
+        token.notifySuccess()
+        return result.getOrNull()!!
+    }
+
+    /**
+     * Handles the termination of the retry loop because of a non-retryable failure by marking the [RetryToken] as
+     * failed and throwing a [RetryFailureException].
+     * @param token The [RetryToken] used in the attempt that was unsuccessful.
+     * @param attempt The number of attempts completed.
+     * @param result The [Result] that yielded the non-retryable condition.
+     */
+    private suspend fun <R> throwFailure(token: RetryToken, attempt: Int, result: Result<R>): Nothing {
+        token.notifyFailure()
+        throw RetryFailureException(
+            "The operation failed",
```

Comment:
suggestion

under the assumption that this exception's message will be visible to customers, perhaps adding some of the kdoc content into the message would be clarifying.  eg "The operation resulted in a non-retryable failure"